File: crabzilla-core/src/main/java/io/github/crabzilla/core/entity/SnapshotPromoter.java
package io.github.crabzilla.core.entity;

import io.github.crabzilla.core.DomainEvent;
import lombok.extern.slf4j.Slf4j;
import lombok.val;

import java.util.List;

@Slf4j
public class SnapshotPromoter<A extends Entity> {

  private final StateTransitionsTrackerFactory<A> trackerFactory;

  public SnapshotPromoter(StateTransitionsTrackerFactory<A> trackerFactory) {
    this.trackerFactory = trackerFactory;
  }

  public Snapshot<A> promote(Snapshot<A> originalSnapshot, Version newVersion, List<DomainEvent> newEvents) {

    if (originalSnapshot.getVersion().getValueAsLong() != newVersion.getValueAsLong() -1) {
      throw new RuntimeException(String.format("Cannot upgrade to version %s since current version is %s",
              newVersion, originalSnapshot.getVersion()));
    }

    val tracker = trackerFactory.apply(originalSnapshot);

    return new Snapshot<>(tracker.applyEvents(c -> newEvents).currentState(), newVersion);
  }

}

File: crabzilla-core/src/main/java/io/github/crabzilla/core/entity/SnapshotPromoter.kt
package io.github.crabzilla.core.entity

import io.github.crabzilla.core.DomainEvent

open class SnapshotPromoter<A : Entity>(private val trackerFactory: (Snapshot<A>) -> StateTransitionsTracker<A>) {

  fun promote(originalSnapshot: Snapshot<A>, newVersion: Version, newEvents: List<DomainEvent>): Snapshot<A> {

    if (originalSnapshot.version.valueAsLong != newVersion.valueAsLong - 1) {
      throw RuntimeException(String.format("Cannot upgrade to version %s since current version is %s",
              newVersion, originalSnapshot.version))
    }

    val tracker: StateTransitionsTracker<A> = trackerFactory.invoke(originalSnapshot)

    return Snapshot(tracker.applyEvents({ newEvents }).currentState(), newVersion)
  }

}
